### PR TITLE
wait for server to spawn before redirecting

### LINKED ISSUE #12: should wait for server to be ready before redirecting
If the notebook server takes a long time to spawn (e.g. because images have to be downloaded), JH will wait for 30 seconds and then redirect to a url like `.../hub/user/...` which fails with a 401 because it's not actually a valid URL (it still gets proxied to the notebooks service for some reason which queries gitlab for a non-existent project, hence the 401). 

The way to fix this is to wait until the server is ready before redirecting. It's unclear how to get this information.

moved from https://github.com/SwissDataScienceCenter/renku/issues/210 -- see https://github.com/SwissDataScienceCenter/renku/issues/210#issuecomment-394503121

### PR DESCRIPTION
**Pull request by rokroskar**
closes #12

### CODE REVIEW

File: src/notebooks_service.py [hunk 1]
```diff
@@ -147,7 +144,32 @@ def launch_notebook(user, namespace, project, commit_sha, notebook=None):
...
 
-    return redirect(notebook_url)
+    # 2. redirect to the server when ready
+    if r.status_code == 201:
+        # server is already running
+        return redirect(notebook_url)
+
+    elif r.status_code == 202:
+        # server is spawning - wait a max of 900 seconds = 15 minutes
+        for i in range(450):
```
Comment (leafty) [inline src/notebooks_service.py]: Is there a way to avoid magic numbers?

Comment (leafty) [inline src/notebooks_service.py]: Is there a way to avoid magic numbers?

Comment (rokroskar) [reply src/notebooks_service.py]: Sure - do you have a suggestion? I'm not sure what an appropriate time to wait would be exactly.

Comment (leafty) [reply src/notebooks_service.py]: What you suggested is fine, it's just that linking 15 minutes to 450*2 happens multiple lines down the comment.
Maybe something like:
```python
timeout = 900 # 900 seconds = 15 minutes
sleep_time = 2
for i in range(timeout / sleep_time):
```

Comment (leafty) [reply src/notebooks_service.py]: You can also use `datetime` objects with a while loop.

Comment (rokroskar) [reply src/notebooks_service.py]: yep, I intended to parametrize this a bit better, I thought your concern was about the length of the timeout itself.

**Pull request merged**